staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,601 @@
+	dependentAPIServices sets.String
+}
+
+var _ DiscoveryAggregationController = &discoveryManager{}
+
+func NewDiscoveryManager(
+	codecs serializer.CodecFactory,
+	serializer runtime.NegotiatedSerializer,
+) DiscoveryAggregationController {
+	return &discoveryManager{
+		serializer:             serializer,
+		mergedDiscoveryHandler: discoveryv1.NewResourceManager(serializer),
+		services:               make(map[serviceKey]*apiServiceInfo),
+		dirtyChannel:           make(chan struct{}),
+	}
+}
+
+func handlerWithUser(handler http.Handler, info user.Info) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		req = req.WithContext(request.WithUser(req.Context(), info))
+		handler.ServeHTTP(w, req)
+	})
+}
+
+// Synchronously refreshes the discovery document by contacting all known
+// APIServices. Waits until all respond or timeout.
+func (dm *discoveryManager) refreshDocument(localOnly bool) error {
+	klog.Info("Refreshing discovery information from apiservices")
+
+	// information needed in the below loop to update a service
+	type serviceUpdateInfo struct {
+		// Service
+		service serviceKey
+
+		// Handler For the service which responds to /discovery/v1 requests
+		handler http.Handler
+
+		// ETag of the existing discovery information known about the service
+		etag string
+	}
+
+	var servicesToUpdate chan serviceUpdateInfo
+
+	// Collect all services which have no discovery document and then update them
+	// Done in two steps like this to avoid holding the lock while fetching the
+	// documents.
+	func() {
+		dm.servicesLock.Lock()
+		defer dm.servicesLock.Unlock()
+
+		servicesToUpdate = make(chan serviceUpdateInfo, len(dm.services))
+
+		// Close the buffered channel once we have finished populating if
+		// Note that even though it is closed it may still deliver any
+		// enqueued items to the webworkers.
+		defer close(servicesToUpdate)
+
+		for key, apiServiceInfo := range dm.services {
+			if !key.isLocalService() && len(apiServiceInfo.dependentAPIServices) == 0 {
+				// Purge unused non-local services from knowledge base
+				delete(dm.services, key)
+			} else if localOnly && !key.isLocalService() {
+				// If only local services are desired to be fetched, skip
+				// non-local
+				continue
+			} else if apiServiceInfo.fresh {
+				// Skip APIServices which are not marked dirty
+				continue
+			}
+
+			// Should not block. Buffered channel is as large as services
+			// slice
+			servicesToUpdate <- serviceUpdateInfo{
+				service: key,
+				handler: apiServiceInfo.handler,
+				etag:    apiServiceInfo.etag,
+			}
+		}
+	}()
+
+	// Download update discovery documents in parallel
+	type resultItem struct {
+		service   serviceKey
+		discovery *metav1.DiscoveryAPIGroupList
+		etag      string
+		error     error
+	}
+
+	waitGroup := sync.WaitGroup{}
+	results := make(chan resultItem, len(servicesToUpdate))
+
+	webworker := func() {
```

Comment:
This doesn't appear to handle the case where the aggregated apiserver only exposes the old discovery doc.  Since we cannot ensure every apiserver updates at the same time, it seems useful to allow for both.  even with an  every-minute load on the original discovery doc, we could bet on it being less than having every client request it.